File: Algorithm/Python/145. 二叉树的后序遍历.py
class Solution:
    def postorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        WHITE, GRAY = 0, 1
        ans = []
        stack = [(WHITE, root)]
        while stack:
            color, node = stack.pop()
            if node == None: continue
            if color == WHITE:
                stack.append((WHITE, node.right))
                stack.append((WHITE, node.left))
                stack.append((GRAY, node))
            else:
                ans.append(node.val)
        return ans

File: Algorithm/C++/145. 二叉树的后序遍历.cpp
class Solution {
public:
    vector<int> postorderTraversal(TreeNode* root) {
        int WHITE = 0, GRAY = 1;
        vector<int> ans;
        stack<pair<int, TreeNode*>> s;
        s.push({WHITE, root});

        TreeNode *node;
        while (!s.empty()){
            int color = s.top().first;
            node = s.top().second;
            s.pop();

            if(node == nullptr){ continue; }
            if(color == WHITE){
                // 需要注意，和递归摆放的位置相反
                s.push({GRAY, node});
                s.push({WHITE, node->right});
                s.push({WHITE, node->left});
            }else{
                ans.push_back(node->val);
            }
        }
        return ans;
    }
};